openedx/license-manager
# feat: send segment events to Kafka message bus

## Description

Send events to the message bus at the same time they are sent to Segment. Eventually these will hopefully replace the direct API calls to enterprise and edx-platform.
This is mostly a "make it work" ticket without too much architectural thought. Ideally, once we have more events live in production, we can make the work setting up producers, serializers, and events more modular and abstract out the actual message bus implementation. 

[ARCHBOM-1984](https://openedx.atlassian.net/browse/ARCHBOM-1984)

## Testing considerations
So far the only way I have been able to manual test this is by adding and deleting subscriptions in Django admin. I have not been able to fire the activate/revoke events locally.

To test:

1.  In devstack, `make dev.up.kafka-control-center && make dev.up.lms`. Kafka sometimes takes a few minutes. You can check by going to the control center at localhost:9021
2. In license-manager/requirements/private.py, set KAFKA_ENABLED=True
3. Bring up license manager (steps are in the repo README)
4. In the LMS Django admin, create an EnterpriseCustomerCatalog (can use an empty EnterpriseCatalogQuery and a fake EnterpriseCustomer, neither will need any real data)
5. In the License Manager Django admin, create a Subscription Plan with some >0 number of licenses using the uuid of the EnterpriseCustomerCatalog. You will also have to create a CustomerAgreement to do this, but that data can all be fake/empty
6. After you create the Subscription Plan, you should see 'license_event' in the Topics tab in the kafka control center
7. Click into 'license_event' > Messages
8. Back in License Manager django admin, delete or add new licenses to the Subscription Plan. Every time you do, you should see a new message appear in the control center. Note that page in the control center only shows messages that have come in since the page loaded so if you refresh, all the messages currently shown will disappear. 

To test on stage:
The only difference between testing on stage and devstack is the cluster used. Stage connects to an actual hosted confluent Kafka cluster. To test on stage, you will need to set 
```
KAFKA_ENABLED=True
KAFKA_BOOTSTRAP_SERVER="pkc-2396y.us-east-1.aws.confluent.cloud:9092"
SCHEMA_REGISTRY_URL="https://psrc-gk071.us-east-2.aws.confluent.cloud"
KAFKA_API_SECRET="<secret>"
SCHEMA_REGISTRY_API_SECRET="<secret>"
KAFKA_API_KEY="<key>"
SCHEMA_REGISTRY_API_KEY="<key>"
LICENSE_TOPIC_NAME="license-event-stage"
```
in the configurations and, in the confluent-cloud dashboard, make sure the api key you are using has CREATE and WRITE permissions for the "license-event-stage" topic.

## Post-review

Squash commits into discrete sets of changes
